aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,126 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
+#include <ATen/Device.h>
 
 namespace at {
 
+// Note [Supervisor deleter]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~
+// SupervisorPtr solves a common problem for allocators of tensor data, which
+// is that the data pointer (e.g., float*) which you are interested in, is not
+// the same as the metadata pointer (e.g., DLManagedTensor) which you need
+// to actually deallocate the data.  Under a conventional deleter design, you
+// have to store extra context in the deleter (the metadata pointer) so that
+// you can actually delete the right thing.  Implementing this with standard
+// C++ is somewhat error-prone: if you use a std::unique_ptr to manage tensors,
+// the deleter will not be called if the data pointer is nullptr, which can
+// cause a leak if the metadata pointer is non-null (and the deleter is
+// responsible for freeing both the data pointer and the metadata pointer).
+//
+// We take a different approach.  The "metadata supervisor" situation is common
+// enough that we have organized our deleter strategy entirely around it:
+// instead of trying to make the deleter for the data pointer handle all the
+// heavy lifting, the data pointer is *non-owning*, and instead there is a
+// (type-erased) supervisor pointer which actually handles deletion.  For simple
+// cases, the supervisor pointer is the same as the data pointer, but if
+// there is some extra metadata, the supervisor pointer points there.
+//
+// There is something of a pattern to writing these; check THAllocator.{h,cpp}
+// for some examples.
+
+using DeleterFnPtr = void(*)(void*);
+using SupervisorPtr = std::unique_ptr<void, DeleterFnPtr>;
+
+// Does not delete anything
+AT_API void deleteNothing(void*);
+// Mints a SupervisorPtr that doesn't do anything.  You must
+// use this, not a nullptr, when you want a view.
+AT_API SupervisorPtr nonOwningSupervisorPtr();
+
+struct SupervisedPtr {
+  // Lifetime tied to supervisor_
+  void* data_;
+  SupervisorPtr supervisor_;
+  Device device_;
+  SupervisedPtr() : data_(nullptr), supervisor_(nonOwningSupervisorPtr()), device_(kCPU) {}
+  SupervisedPtr(void* data, SupervisorPtr&& supervisor, Device device)
+    : data_(data), supervisor_(std::move(supervisor)), device_(device) {}
+  void* operator->() const { return data_; }
+  void* get() const { return data_; }
+  operator bool() { return data_ || supervisor_; }
```

Comment:
do we need this (and all the operator== below) looking at the supervisor?  This has surprising behavior, like
`
SupervisedPtr x;
bool(x)
`
evaluating to true.

Reply:
Same explanation as above.